lms/djangoapps/email_marketing/signals.py

```diff
@@ -54,41 +55,33 @@ def add_email_marketing_cookies(sender, response=None, user=None,
         if sailthru_content:
             post_parms['cookies'] = {'anonymous_interest': sailthru_content}
 
-    try:
-        sailthru_client = SailthruClient(email_config.sailthru_key, email_config.sailthru_secret)
-        log.info(
-            'Sending to Sailthru the user interest cookie [%s] for user [%s]',
-            post_parms.get('cookies', ''),
-            user.email
-        )
-        time_before_call = datetime.datetime.now()
+    time_before_call = datetime.datetime.now()
+    sailthru_response = get_email_cookies_via_sailthru.delay(user.email, post_parms)
 
-        sailthru_response = \
-            sailthru_client.api_post("user", post_parms)
+    try:
+        # synchronous call to celery task with a timeout limit
+        sailthru_response.get(timeout=email_config.user_registration_cookie_timeout_delay,
+                              propagate=True)
+        cookie = sailthru_response.result
 
+    except TimeoutError as exc:
+        log.error("Timeout error while attempting to obtain cookie from Sailthru: %s", unicode(exc))
```

Comment:
Okay, I think I see what we're getting at here. The above `delay` call is executing the task asynchronously, then the following `get` will block until it finishes and has a result to give you. This is **not** the same thing as calling a celery task synchronously in-process. Take a look at the code celery uses [here](https://github.com/celery/celery/blob/3.1/celery/app/task.py), and note that `apply_async` calls `send_task`, the piece that sends a message over rabbitmq to be picked up by a worker. `apply`, on the other hand, will set up the task with all the celery logging you'd expect, but skip rabbitmq entirely and execute the task in-process.

That said, it's hard/impossible to do what you're attempting all in one process - *something* has to keep track of the timeout and interrupt if it expires. I think the approach you've taken here will work with that.

> When i use threads to simulate multiuser and invoke tasks .. all results in timeout error, but when i schedule them one by one and wait before next calls or in thread case use locks .. these blocking task runs fine. Iam not sure what would happen on the case when multiple django processes would be populating rabbitmq for a task, would they give time out error as iam getting. 

How/where are you doing this simulation? If it locally on your devstack, you'll need to use multiple processes with [--setting=devstack_with_worker](https://github.com/edx/edx-platform/blob/master/lms/envs/devstack_with_worker.py), otherwise the initial `delay` gets converted to a synchronous call, run in-process (Because envs.common.py sets [CELERY_ALWAYS_EAGER](http://docs.celeryproject.org/en/3.1/configuration.html)).

It's important to note that "blocking" in this context means "the process that called get() is blocked until a result is returned" not "only one process can execute this task at a time". In production, it'd be very likely that there would be multiple workers running `get_email_cookies_via_sailthru` at the same time, with multiple main processes waiting on the results independently. This should work fine, but be aware of it.

Overall, I think this will work out as you expect, there's nothing that looks inherently incorrect. The ideal fix would be for `sailthru_client.api_post` to take a timeout param, but that seems out of our reach. LMK if I can assist with your testing.